state/protocol/protocol_state/state/mutable_protocol_state_test.go

```diff
@@ -0,0 +1,650 @@
+		orderedServiceEvents = append(orderedServiceEvents, sealedResult.ServiceEvents...)
+		seals = append(seals, seal)
+
+		s.headersDB.On("ByBlockID", sealedBlock.ID()).Return(sealedBlock, nil)
+		s.resultsDB.On("ByID", seal.ResultID).Return(sealedResult, nil)
+	}
+	slices.Reverse(seals) // revert order of seals, thereby they are listed in order of _decreasing_ block height
+
+	s.Run("service events leave state invariant", func() {
+		parentProtocolStateID := s.parentState.ID()
+		s.evolvingState.On("ID").Return(parentProtocolStateID, nil).Once()
+
+		s.kvStateMachines[0] = s.mockStateTransition().ExpectedServiceEvents(orderedServiceEvents).Mock()
+		s.kvStateMachines[1] = s.mockStateTransition().ExpectedServiceEvents(orderedServiceEvents).Mock()
+
+		s.testEvolveState(seals, parentProtocolStateID, false)
+	})
+
+	s.Run("service events change state", func() {
+		expectedResultingStateID := unittest.IdentifierFixture()
+		modifyState := func(_ mock.Arguments) {
+			s.evolvingState.On("ID").Return(expectedResultingStateID, nil).Once()
+		}
+		s.kvStateMachines[0] = s.mockStateTransition().ExpectedServiceEvents(orderedServiceEvents).DuringEvolveState(modifyState).Mock()
+		s.kvStateMachines[1] = s.mockStateTransition().ExpectedServiceEvents(orderedServiceEvents).Mock()
+
+		s.testEvolveState(seals, expectedResultingStateID, true)
+	})
+
+}
+
+// Test_ParentNotFound checks the behaviour of `MutableProtocolState.EvolveState` when the specified parent block is not found.
+func (s *StateMutatorSuite) Test_InvalidParent() {
+	unknownParent := unittest.IdentifierFixture()
+	s.protocolKVStoreDB.On("ByBlockID", unknownParent).Return(nil, storage.ErrNotFound)
+
+	_, dbUpdates, err := s.mutableState.EvolveState(unknownParent, s.candidate.View, []*flow.Seal{})
+	require.Error(s.T(), err)
+	require.True(s.T(), dbUpdates.IsEmpty())
+}
+
+// Test_ReplicateFails verifies that errors during the parent state replication are escalated to the caller.
+// Because the failure is arising early, we don't expect calls to the state machine constructors to be called
+// (they require a replica of the parent state as target, which is not available if replication fails)
+func (s *StateMutatorSuite) Test_ReplicateFails() {
+	exception := errors.New("exception")
+	s.parentState = *protocol_statemock.NewKVStoreAPI(s.T())
+	s.parentState.On("GetProtocolStateVersion").Return(s.latestProtocolVersion)
+	s.parentState.On("GetVersionUpgrade").Return(nil).Once()
+	s.parentState.On("Replicate", s.latestProtocolVersion).Return(nil, exception).Once()
+
+	// `SetupTest` initializes the mock factories to expect to be called, so we overwrite the mocks here:
+	s.kvStateMachineFactories[0] = *protocol_statemock.NewKeyValueStoreStateMachineFactory(s.T())
+	s.kvStateMachineFactories[1] = *protocol_statemock.NewKeyValueStoreStateMachineFactory(s.T())
+
+	_, dbUpdates, err := s.mutableState.EvolveState(s.candidate.ParentID, s.candidate.View, []*flow.Seal{})
+	require.Error(s.T(), err)
```

Comment:
```suggestion
	require.ErrorIs(s.T(), err, exception)
```

Reply:
done in https://github.com/onflow/flow-go/pull/5644/commits/9c38cfa2bb5b0d4b2e0b2dc4dd6e78c548ff2cdb